### PR TITLE
Squad

### PR DESCRIPTION
**Pull request by yzhang123**
Adding Code for Question Answering with Squad v1.1 and v2.0,
unit tests, and tutorial.

### CODE REVIEW

File: collections/nemo_nlp/nemo_nlp/data/data_layers.py [hunk 1]
```diff
@@ -733,6 +735,86 @@ def __init__(self,
         super().__init__(dataset_type, dataset_params, **kwargs)
 
 
+class BertQuestionAnsweringDataLayer(TextDataLayer):
```
Comment (blisc) [inline collections/nemo_nlp/nemo_nlp/data/data_layers.py]: Please add to your docstring what each of the arguments for the initializer means.

File: collections/nemo_nlp/nemo_nlp/data/datasets/squad.py [hunk 2]
```diff
@@ -0,0 +1,846 @@
...
+import string
+
+import numpy as np
+from tqdm import tqdm
+
+from torch.utils.data import Dataset
+import torch
+from nemo.utils.exp_logging import get_logger
+from .utils import DataProcessor
+from ...utils.metrics.squad_metrics import _compute_softmax, \
```
Comment (blisc) [inline collections/nemo_nlp/nemo_nlp/data/datasets/squad.py]: please do import (
  item1
  item2
  ...
)

File: collections/nemo_nlp/nemo_nlp/data/datasets/squad.py [hunk 3]
```diff
@@ -0,0 +1,846 @@
...
+
+
+"""
+Utility functions for Question Answering NLP tasks
+Some parts of this code were adapted from the HuggingFace library at
+https://github.com/huggingface/transformers
+"""
+
+
+class SquadDataset(Dataset):
```
Comment (blisc) [inline collections/nemo_nlp/nemo_nlp/data/datasets/squad.py]: Class needs a dosctring

File: collections/nemo_nlp/nemo_nlp/data/datasets/squad.py [hunk 4]
```diff
@@ -0,0 +1,846 @@
...
+    def __getitem__(self, idx):
+        feature = self.features[idx]
+        return (np.array(feature.input_ids),
+                np.array(feature.segment_ids),
+                np.array(feature.input_mask),
+                np.array(feature.start_position),
+                np.array(feature.end_position),
+                np.array(feature.unique_id))
+
+    def get_predictions(self, unique_ids, start_logits, end_logits,
```
Comment (blisc) [inline collections/nemo_nlp/nemo_nlp/data/datasets/squad.py]: Please have one argument per line

File: collections/nemo_nlp/nemo_nlp/data/datasets/squad.py [hunk 5]
```diff
@@ -0,0 +1,846 @@
...
+
+            prediction = preds[qas_id]
+            exact_scores[qas_id] = \
+                max(exact_match_score(a, prediction) for a in gold_answers)
+            f1_scores[qas_id] = \
+                max(f1_score(a, prediction) for a in gold_answers)
+
+        return exact_scores, f1_scores
+
+    def evaluate(self, unique_ids, start_logits,
```
Comment (blisc) [inline collections/nemo_nlp/nemo_nlp/data/datasets/squad.py]: Please have one argument per line

File: collections/nemo_nlp/nemo_nlp/data/datasets/squad.py [hunk 6]
```diff
@@ -0,0 +1,846 @@
...
+                                 max_answer_length, do_lower_case,
+                                 version_2_with_negative,
+                                 null_score_diff_threshold)
+
+        exact_match, f1 = self.evaluate_predictions(all_predictions)
+
+        return exact_match, f1, all_predictions
+
+
+def convert_examples_to_features(examples, tokenizer, max_seq_length,
```
Comment (blisc) [inline collections/nemo_nlp/nemo_nlp/data/datasets/squad.py]: Please have one argument per line

File: collections/nemo_nlp/nemo_nlp/utils/callbacks/squad.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+        if 'unique_ids' in kv:
+            unique_ids = []
+            for v_tensor in v:
+                for id_tensor in v_tensor:
+                    unique_ids.append(id_tensor.detach().cpu().tolist())
+            global_vars['eval_unique_ids'].extend(unique_ids)
+
+
+def eval_epochs_done_callback(global_vars, eval_data_layer, do_lower_case,
```
Comment (blisc) [inline collections/nemo_nlp/nemo_nlp/utils/callbacks/squad.py]: Please have one argument per line

File: examples/nlp/squad.py [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+Model	                EM      F1
+BERT Based uncased      80.59    88.34
+BERT Large uncased      83.88    90.65
+"""
+
+import argparse
+import os
+import sys
+import json
+import nemo
```
Comment (blisc) [inline examples/nlp/squad.py]: Can you group nemo and nemo_nlp imports?
Also a huge plus if your imports are alphabetical within a section. Eg all python-native libraries together in alphabetical order, then all pip packages in alphabetical order, then nemo in alphabetical order

File: examples/nlp/squad.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+import nemo_nlp
+from nemo.utils.lr_policies import get_lr_policy
+from nemo_nlp import BertQuestionAnsweringDataLayer
+from nemo_nlp import NemoBertTokenizer, SentencePieceTokenizer
+from nemo_nlp import QuestionAnsweringLoss
+from nemo_nlp.utils.callbacks.squad import \
+    eval_iter_callback, eval_epochs_done_callback
+
+parser = argparse.ArgumentParser(description="Squad_with_pretrained_BERT")
```
Comment (blisc) [inline examples/nlp/squad.py]: Can you wrap the code inside a if __name__ == "__main__"?

File: examples/nlp/squad.py [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+                    mode=mode,
+                    version_2_with_negative=version_2_with_negative,
+                    batch_size=batch_size,
+                    tokenizer=tokenizer,
+                    data_dir=args.data_dir,
+                    max_query_length=max_query_length,
+                    max_seq_length=max_seq_length,
+                    doc_stride=doc_stride)
+
+    input_ids, input_type_ids, input_mask, \
```
Comment (blisc) [inline examples/nlp/squad.py]: Can you reformat this as per #268

File: tests/test_squad.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
+# Copyright (c) 2019 NVIDIA Corporation
+import unittest
+import os
+import sys
+import json
+
+import shutil
```
Comment (blisc) [inline tests/test_squad.py]: Please group this with the rest of the Standard library imports.

File: tests/test_squad.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+            download_script_path = os.path.abspath(
+                os.path.join(os.path.dirname(__file__),
+                             '../examples/nlp/scripts'))
+            sys.path.insert(0, download_script_path)
+            from download_squad import SquadDownloader
+            print("Extracting Squad data to: {0}".format(squad_folder))
+            squad_dl = SquadDownloader(data_folder)
+            squad_dl.download()
+
+            squad_v1_dev_file = \
```
Comment (blisc) [inline tests/test_squad.py]: How big is this dataset? If it is big, maybe we should move it to CI instead of unittests.

Comment (yzhang123) [reply tests/test_squad.py]: its only part of the dataset. Both tests take 100s

File: collections/nemo_nlp/nemo_nlp/data/datasets/squad.py (hunk 3 above)
Comment (yzhang123) [reply collections/nemo_nlp/nemo_nlp/data/datasets/squad.py]: I changed the source branch accordingly

File: collections/nemo_nlp/nemo_nlp/data/datasets/squad.py (hunk 4 above)
Comment (yzhang123) [reply collections/nemo_nlp/nemo_nlp/data/datasets/squad.py]: I changed the source branch accordingly

File: examples/nlp/squad.py (hunk 8 above)
Comment (yzhang123) [reply examples/nlp/squad.py]: I changed the source branch accordingly

File: examples/nlp/squad.py (hunk 10 above)
Comment (yzhang123) [reply examples/nlp/squad.py]: I changed the source branch accordingly

File: tests/test_squad.py (hunk 11 above)
Comment (yzhang123) [reply tests/test_squad.py]: I changed the source branch accordingly

File: collections/nemo_nlp/nemo_nlp/utils/callbacks/squad.py (hunk 7 above)
Comment (blisc) [reply collections/nemo_nlp/nemo_nlp/utils/callbacks/squad.py]: I don't think this has been done yet

File: examples/nlp/scripts/download_squad.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import urllib.request
+import argparse
```
Comment (blisc) [inline examples/nlp/scripts/download_squad.py]: Can you put these in alphabetical order?

File: collections/nemo_nlp/nemo_nlp/utils/metrics/squad_metrics.py [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+
+import math
+import collections
+import re
+import string
```
Comment (blisc) [inline collections/nemo_nlp/nemo_nlp/utils/metrics/squad_metrics.py]: Can you put these four in alphabetical order?
transformers import should still stay below these four

File: collections/nemo_nlp/nemo_nlp/data/datasets/squad.py [hunk 15]
```diff
@@ -0,0 +1,903 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import os
+import sys
+import json
+import collections
+import pickle
+import string
```
Comment (blisc) [inline collections/nemo_nlp/nemo_nlp/data/datasets/squad.py]: Can you clean up these imports too?

File: collections/nemo_nlp/nemo_nlp/data/data_layers.py [hunk 16]
```diff
@@ -17,20 +17,22 @@
            'BertPretrainingPreprocessedDataLayer',
            'TranslationDataLayer',
            'GlueDataLayerClassification',
-           'GlueDataLayerRegression']
+           'GlueDataLayerRegression',
+           'BertQuestionAnsweringDataLayer']
 
 # from abc import abstractmethod
 import sys
```
Comment (blisc) [inline collections/nemo_nlp/nemo_nlp/data/data_layers.py]: Can you clean up these imports too?

Comment (blisc) [reply collections/nemo_nlp/nemo_nlp/data/data_layers.py]:
```
import sys
import os
import random

import numpy as np
import h5py
import torch
from torch.utils import data as pt_data
```
Can you make these in alphabetical order (os, random, sys) and (h5py, numpy, torch)?

**Comment by blisc** [comment]
> This pull request **introduces 4 alerts** when merging [3bdbde5](https://github.com/NVIDIA/NeMo/commit/3bdbde56f1f234162cbd870f4e5b177842c19804) into [2bb07bc](https://github.com/NVIDIA/NeMo/commit/2bb07bcda2e3e247efb1c7556560f05cf2bf0e65) - [view on LGTM.com](https://lgtm.com/projects/g/NVIDIA/NeMo/rev/pr-a420ebacc0dd4dc2c307b04ec4279b357f94e30b)
> 
> **new alerts:**
> 
> * 2 for Unused import
> * 1 for Signature mismatch in overriding method
> * 1 for Module is imported with 'import' and 'import from'

Can you remove the unused imports?

**Pull request merged**